### ISSUE TITLE
Pangeo + CEDA / JASMIN

### ISSUE THREAD

**Issue by rabernat** [pangeo-data/pangeo#396]
The UK NERC [Center for Environmental Data Analysis](http://www.ceda.ac.uk/) has a mission closely aligned with Pangeo.

> The Centre for Environmental Data Analysis (CEDA) serves the environmental science community through three data centres, data analysis environments, and participation in a host of relevant research projects.
>
> We aim to support environmental science, further environmental data archival practices, and develop and deploy new technologies to enhance access to data. Additionally we provide services to aid large scale data analysis.

In particular, they operate the [JASMIN](http://www.ceda.ac.uk/analysis-environments/) service. Has a Pangeo environment been deployed here? Is this something anyone wants to try?

I am randomly tagging some CEDA people to see if they are interested in discussing: @mkjpryor-stfc, @bnlawrence, @duncanwp, @agstephens. Are there some ways Pangeo could help with your mission?

**Comment by mrocklin** [comment]
We worked closely with CEDA when developing the dask-kubernetes project.  That work came from early work by @mkjpryor-stfc and others.

**Comment by philipkershaw** [comment]
Hi @rabernat, I'm tech manager for CEDA.   We're definitely interested in collaboration with Pangeo and we have been tracking activities since @kmpaul first told me about the project.   We have a project the NERC DataLab which could be of interest.  It provides notebooks and integrates with Dask and PySpark.  It deploys on Kubernetes.  We have run it in a project on our OpenStack cloud.  @mkjpryor-stfc has been closely involved technically and could provide more of the details.   We are also involved in ESGF and have the face-to-face meeting in December in Washington DC.  There was some interest from the collaboration in inviting someone to speak about Pangeo at the meeting so someone may have been in touch about this.

**Comment by rabernat** [comment]
Thanks for for the info and sorry for the slow follow up!

Pangeo is not only focused on cloud / kubernetes style environments--we are also able to deploy on standard HPC systems with no special treatment. This is what we do today on Cheyenne (see docs: http://pangeo.io/setup_guides/hpc.html).

So it should be possible to run pangeo on Jasmine today purely from user space, without any special privileges. I am wondering if anyone would like to give this a go and see how it works. If so, we could include some documentation about how to do this.

**Comment by duncanwp** [comment]
I'd definitely be up for trying this out. I tried to make this work on the ARCHER data processing facility (https://www.archer.ac.uk/documentation/rdf-guide/) a while ago but couldn't quite get the client to talk to the workers. That guide looks very helpful though.

I'll give it a go and report back!

**Comment by duncanwp** [comment]
So I have a notebook tunnelled through to the server, a client and cluster setup, but I can't seem to get the client to see the workers (which is a similar problem I was having on ARCHER).

I'm using the following cluster setup:

``` 
from dask_jobqueue import LSFCluster

cluster = LSFCluster(cores=16, memory='128 GB',  # The size of the ivybridge128G nodes
                     queue='par-multi',
                     processes=16, # For memory intensive jobs you should reduce this number
                     job_extra=['-R "select[maxmem > 96000] rusage[mem=96000]"'],  # Be sure to ask for more memory,
                     walltime='2:00')
```
Which when I start workers gives me a sensible output:

```
DEBUG:dask_jobqueue.core:starting 8 workers
DEBUG:dask_jobqueue.core:writing job script: 
#!/bin/bash

#BSUB -J dask-worker
#BSUB -e dask-worker.err
#BSUB -o dask-worker.out
#BSUB -q par-multi
#BSUB -n 16
#BSUB -M 128000
#BSUB -W 2:00
#BSUB -R "select[maxmem > 96000] rusage[mem=96000]"
JOB_ID=${LSB_JOBID%.*}



/home/users/dwatsonparris/miniconda3/envs/pangeo/bin/python -m distributed.cli.dask_worker tcp://130.246.142.77:44963 --nthreads 1 --nprocs 16 --memory-limit 8.00GB --name dask-worker--${JOB_ID}-- --death-timeout 60

DEBUG:dask_jobqueue.core:Submitting the following calls to command line
DEBUG:dask_jobqueue.core:bsub < /tmp/tmpck_pdbvi.sh 2> /dev/null
DEBUG:dask_jobqueue.core:started job: 3198698
```

But neither the Client nor the Dask dashboard can see the resulting workers (even though the job is running):

<img width="1077" alt="screenshot 2018-10-05 at 15 20 34" src="https://user-images.githubusercontent.com/3169620/46541100-0433b980-c8b3-11e8-9cb2-34f9785b13ed.png">
<img width="1377" alt="screenshot 2018-10-05 at 15 20 07" src="https://user-images.githubusercontent.com/3169620/46541107-09910400-c8b3-11e8-8808-ce947668c634.png">

I've attached the job [output](https://github.com/pangeo-data/pangeo/files/2450692/dask-worker.out.txt) and [error](https://github.com/pangeo-data/pangeo/files/2450691/dask-worker.err.txt) files, but they don't show anything more than warnings. I'm not sure how to debug this further so any thoughts are welcome!

**Comment by mrocklin** [comment]
Looks like your workers were never able to contact your scheduler.  My
guess is that this address isn't visible from where your workers have been
created.

    distributed.worker - INFO - Waiting to connect to: tcp://
130.246.142.77:44963

**Comment by duncanwp** [comment]
> My guess is that this address isn't visible from where your workers have been created.

Yes, I suspect you're right. I've emailed the support team to understand if this is a fundamental limitation of the setup they have.

In the meantime are there any other mechanisms to allow this communication? I seem to remember in iPython you can use ssh/file based communication between the scheduler and the workers.

**Comment by mrocklin** [comment]
Dask only uses TCP for inter-node communication.  Probably you just need to
find and select the right network interface.

https://stackoverflow.com/a/43881158/616616

**Comment by mkjpryor-stfc** [comment]
I'd be very surprised if that port is open in the firewall config. I also think you'll have a hard time convincing the infrastructure team to open it. We've discussed running Jupyter notebooks on LOTUS before and the response is never positive. Having said that, it would be nice to be able to use Dask...

I think using the cloud is going to be the best option for getting Pangeo running on JASMIN in the first instance.

**Comment by duncanwp** [comment]
Also, just to be clear, this is only between jasmin-sci1 and the lotus compute nodes. So inside the external firewall.

**Comment by mkjpryor-stfc** [comment]
@mrocklin What I said pretty much applies to any port. The operations team are reluctant to open ports that aren't for system services.

Presumably this would have to be a range that was opened anyway? To account for the case where two user's Dask schedulers get scheduled (by LSF) to the same node?

Is the Dask transport pluggable? I can't believe we are the only people who would prefer an MPI transport...

**Comment by mkjpryor-stfc** [comment]
@duncanwp So your notebook server is running on jasmin-sci1? How are you connecting your local browser to it? Using an SSH tunnel?

**Comment by mrocklin** [comment]
> Is the Dask transport pluggable? I can't believe we are the only people who would prefer an MPI transport...

Yes.  https://distributed.readthedocs.io/en/latest/communications.html#communications-api
An MPI implementation would be welcome.  See also https://github.com/dask/distributed/issues/2177

**Comment by duncanwp** [comment]
@mkjpryor-stfc Yes, exactly

**Comment by mrocklin** [comment]
> The operations team are reluctant to open ports that aren't for system services.

What I typically see is that people run the dask scheduler and workers within an internal network, where every machine can talk to every other machine on any port, but can't see outside.  Usually this means that you have to specify an internal network interface using the `--interface` keyword.

They might also run a Jupyter notebook within that network.

They then use ssh tunnelling to get access to the Jupyter port outside the network to their local machine.

**Comment by philipkershaw** [comment]
I know that Dask has been used with Lotus on JASMIN as I spoke to a user about it recently.  I can write to them and find out more if it would be useful

**Comment by duncanwp** [comment]
@philipkershaw Yes, that would be great, thanks.

I've tried specifying the only other network interface and while it's definitely trying to connect on a different subnet I have the same problem.

**Comment by rabernat** [comment]
We have done this successfully on many different HPC systems, including some with very high security requirements such as NASA’s Pleiades computer.

We are not talking about opening ports to the outside world. We are talking about TCP communication between different compute nodes within the cluster. I have never encountered an HPC system where such communication were prohibited. Surely dask is not the only application that might want to use TCP on Jasmine.

**Comment by duncanwp** [comment]
I think, to be fair to CEDA, the architecture of JASMIN plays a role here: The LOTUS compute cluster is (I believe) on a separate network to the interactive science VMs from which I'm running the scheduler. If I was able to launch an interactive session on LOTUS from the science VMs to run the scheduler it might be easier. Hopefully we can figure a way through this though.

**Comment by mrocklin** [comment]
Ideally you would launch your scheduler on the same network where you launch your workers

**Comment by philipkershaw** [comment]
@duncanwp, I heard back from the user - Jose at UCL.  He shared these notes: https://paper.dropbox.com/doc/Dask-on-CEMSJASMINWhatever-pJv2kvR2oPhehee7zQ4C1

**Comment by duncanwp** [comment]
Yeah, so this is one approach but it’s not very satisfactory and there are easier ways of doing it. 

It’s fine for a parallelising a script but by submitting a separate job for the scheduler you have to manage two jobs manually and lose any interactivity with the workers. 

I’ve raised a support ticket for allowing the workers to communicate but perhaps there are other channels for raising this use case with the infrastructure team?

**Comment by guillaumeeb** [comment]
I'm not sure if you've tried the usual solution of submiting a job, be it interactive or not, for running your notebook  on the compute node. This way you should be able to create a dask cluster using dask-jobqueue inside the notebook.

**Comment by guillaumeeb** [comment]
See maybe http://jobqueue.dask.org/en/latest/#using-jupyter.

**Comment by mkjpryor-stfc** [comment]
@duncanwp I suspect running the notebook server on the sci machines will not be popular with the operations team... It is not scalable at all as every user would need their own port.

**Comment by mrocklin** [comment]
I often see people ssh in and use ssh tunnelling to get the notebook port
ot without getting permission from IT.  Is ssh accessible to most users?

**Comment by mkjpryor-stfc** [comment]
@mrocklin I think that is what @duncanwp is doing at the moment, which works. He is running the notebook server and Dask scheduler on a shared machine that is separate to the batch cluster, and connecting to the notebook server using an SSH tunnel. He is then attempting to launch Dask workers in the cluster, but they can't contact the Dask scheduler because the port it is using is not open on the shared machine.

This is not really the expected use for the shared machines, and I suspect our operations team would prefer to find a better way to do it.

**Comment by duncanwp** [comment]
After a little more experimenting I can just about make this work by running the notebook in a non-interactive job, figuring out which host it’s being run on and the tunnelling to that. It’s not elegant, and definitely not something I’d recommend to other users though since it involves three ssh tunnels (through login1, Jasmin-sci1 then to the lotus host), each of which could introduce a port clash. 

@mkjpryor-stfc I’d be very happy to be involved with a telecom at some point to try and come to a pragmatic solution. Perhaps allowing interactive jobs on lotus from the login node? Or just opening up the ports between the compute nodes. 

I think this could be a really popular use case once it’s working since a large proportion of the compute on lotus are the kind of big data operations dask makes easy. It would be a shame to restrict it in search of a perfect solution!

**Comment by rabernat** [comment]
> I suspect running the notebook server / Dask scheduler on the sci machines will not be popular with the operations team... It is not scalable at all as every user would need their own port.

Are the "sci machines" shared? If not, there should not be any problems.

**Comment by mkjpryor-stfc** [comment]
@rabernat The sci machines are shared. We do have an equivalent that is not shared, which is a cloud tenancy that lives inside the firewall (users do not get root). This could be an option for the scheduler/notebook server.

**Comment by rabernat** [comment]
Ok, I think I finally understand why Jasmin is different from other clusters. On NCAR, NASA, NSF HPC systems, we are generally dealing with non-shared compute nodes. In this case, it's straightforward for a user to start services (jupyter notebook, dask scheduler) on their compute node and connect to them via ssh tunnel (for the notebook) / inter-node TCP communication (for the dask workers).

**Comment by mkjpryor-stfc** [comment]
@rabernat The TCP communication is an orthogonal problem. Our infrastructure guys are way stricter on network security, even on the internal network, than most other HPC centres. I think it has to do with the structure of the network and connections with other systems - JASMIN doesn't have a totally isolated network like a lot of HPC machines.

**Comment by philipkershaw** [comment]
@mkjpryor-stfc, I think we should look at this -
> a cloud tenancy that lives inside the firewall (users do not get root). This could be an option for the scheduler/notebook server.

Perhaps we can do something ready or semi-ready made along the lines of the current <opendap / group workspace access>-as-as-service

**Comment by rabernat** [comment]
This new blog post that @jhamman just published is highly relevant to the discussion here:
https://medium.com/pangeo/dask-jobqueue-d7754e42ca53

It describes some of the challenges of running dask on hpc and the approach we are trying to encourage using dask-jobqueue.

**Comment by mkjpryor-stfc** [comment]
@rabernat 

Interesting read.

I hate to bang on about this, but I think this article also highlights why an MPI Comm implementation would be useful, and not only because it would avoid having to open unusual ports for TCP traffic.

Every HPC system in the world is tuned for MPI performance. Things like having to specify the Inifiniband network interface manually (which could easily be missed) would be removed, because the tuned MPI implementation would take care of that.

I have expertise with Python and MPI (though not in Python), so could be a good candidate to code this up, but I have a lot of pressure on my time currently so not sure when I could have a crack at it.

**Comment by duncanwp** [comment]
I'd also be interested in helping out with this. I'm not clear what really is required though, could the distributed.Comm object just wrap mpi4py, or would it need to be lower level?

I'm also not sure how cleanly the distributed.Comm model of distinct objects for each communication channel will map to MPI where one 'comm_world' manages communication between all workers.

**Comment by mrocklin** [comment]
An MPI comm would be a welcome contribution.  

> could the distributed.Comm object just wrap mpi4py, or would it need to be lower level?

Yes, using mpi4py is probably best.

> I'm also not sure how cleanly the distributed.Comm model of distinct objects for each communication channel will map to MPI where one 'comm_world' manages communication between all workers.

Yeah, my guess is that MPI wouldn't work well if you wanted to add new workers mid-computation or if workers died during execution (there's no obvious "other end closed communication" signal in MPI).  Assuming you're staying in a well protected computation though things should be ok.

I do think that TCP on Infiniband is more likely to be useful to people, but IT's preference for MPI is something to consider.  In either event there's no reason not to have both.

**Comment by mrocklin** [comment]
> and not only because it would avoid having to open unusual ports for TCP traffic.

Just to be clear, most HPC networks, even *very* secure ones tend to leave all ports open internally.  Yours is the first that I've seen (out of maybe a hundred or so across academy, university, and government) that don't allow worker nodes to talk to each other.

**Comment by mkjpryor-stfc** [comment]
@mrocklin 

Having thought some more, it doesn't seem to be specifically TCP that is the problem. @duncanwp was able to get it working when he ran the scheduler *inside* LOTUS (the batch cluster) using a job. It doesn't work if the scheduler is on one of the shared interactive VMs, where the port restrictions are greater (probably rightly so). As we've said previously, I'm nervous about the use of shared systems for the notebook server / scheduler.

**Comment by mrocklin** [comment]
> It doesn't work if the scheduler is on one of the shared interactive VMs, where the port restrictions are greater (probably rightly so). As we've said previously, I'm nervous about the use of shared systems for the notebook server / scheduler.

Yeah, this is pretty common.  The general solution is to run everything within the cluster and use SSH tunnelling to get access to Jupyter.  Usually IT is fine with exposing ssh access.  (I think we're now repeating our conversation from above)

**Comment by mkjpryor-stfc** [comment]
@mrocklin 

I think we are repeating. @duncanwp - can you document the process of running and accessing the whole lot using jobs? As I said before, LOTUS should _in theory_ support interactive jobs, which might be better for the notebook server. Fatima should know more.

**Comment by duncanwp** [comment]
@mkjpryor-stfc Here's a quick [write-up ](https://paper.dropbox.com/doc/Dask-jobqueue-on-JASMIN--APCst4N3A3BNtvqTC6KfltfDAQ-1xc6gq5LJuRKZ0WE3O1MX) of my attempt and my take on the difficulties and possible solutions. Happy to discuss offline if it would be helpful.

**Comment by rabernat** [comment]
Given the upcoming drive for CMIP6, is it worth reopening this issue to discuss whether we can get Pangeo running on Jasmin? There will be many PB of CMIP6 data coming to Jasmin very soon. Pangeo can help greatly with this.

@duncanwp's excellent writeup explains what needs to be done to get this to work. Is this solution considered sufficient? In my reading, it seems like there are a few things that could be done on the system side to smooth the process.

Any interest from the CEDA folks in providing support for Pangeo-style analysis going forward?

**Comment by philipkershaw** [comment]
Hi @rabernat, noted it would be good to follow up with @duncanwp and the rest of the JASMIN team about this.  That said, we hope be to deploying a Jupyter service on our OpenStack cloud soon with access to data from the CEDA archive (CMIP6 included).  Using this cloud service may be an easier route for the moment than through the batch compute environment.

Colleague @mkjpryor-stfc has also worked on a project to create ready made Pangeo instances - amongst other things on our cloud - https://github.com/cedadev/jasmin-cluster-as-a-service/wiki   This is in test but again we plan to put into production later in the year.

**Comment by rsignell-usgs** [comment]
In case it's relevant, Andrea Zonca (@zonca) has been working on getting Pangeo going on NSF's XSEDE Jetstream cloud, which uses OpenStack.   He has a number of [blog posts](https://zonca.github.io/) on his progress.

**Comment by rabernat** [comment]
Reopening this to find out what is the status of Pangeo on JASMIN.

**Comment by duncanwp** [comment]
We've had some offline discussions and arranged for a small group to test a cluster-as-a-service approach. It's not quite the flexibility it sounds like other centers give, but should allow us to spin-up Pangeo and still (hopefully) have access to their data archive.

**Comment by rabernat** [comment]
Great! [Dask gateway](https://github.com/dask/dask-gateway) has come a long way since we last talked, and this could be a great solution.

**Comment by duncanwp** [comment]
Nice, I'll check it out! Thanks

**Comment by philipkershaw** [comment]
@rabernat, there's an article here about the JASMIN Cluster-as-a-Service:

https://help.jasmin.ac.uk/article/4735-cluster-as-a-service

CaaS is a wider service for JASMIN's OpenStack cloud to enable users to deploy different preassembled components such as a Slurm cluster or baseline Kubernetes from a point and click interface.  Amongst those options is one to deploy Pangeo.  From a Pangeo instance deployed there, access to our archive and users' group workspaces is available over OPeNDAP.  We are also working on a link up to enable direct POSIX access to our data archive.  We also have a new object store with S3 interface which some users are starting to use with netCDF datasets

**Comment by wseviour** [comment]
I'm part of the team organising a UK CMIP6 hackathon later this year, and I'm just reopening this to see if there have been any changes with Pangeo on JASMIN?  In particular, I'd be interested in any updates on the ability to use `dask_jobqueue` and the Dask dashboard with the LOTUS HPC cluster, following on from @duncanwp's issue above. I see there has been a transition from LSF to SLURM over the last year, so perhaps permissions have changed along with that?

**Comment by rabernat** [comment]
My understanding is that Pangeo support on CEDA has evolved a lot. I'll ping my new friend @agstephens who may have a n update.

**Comment by agstephens** [comment]
@wseviour There have indeed been various updates over the last couple of years, and we are aware of the plans for the UK COP26 Hackathons. Please contact me for more info directly on: ag.stephens@stfc.ac.uk

**Comment by rabernat** [comment]
Thanks for the reply Ag!

In addition to contacting you personally by email, is there any public website where users can learn about Pangeo support on CEDA? Is this page up to date: https://help.jasmin.ac.uk/article/4797-cluster-as-a-service-pangeo ?

**Comment by agstephens** [comment]
Hi @rabernat, that page should be up-to-date - but it doesn't yet have anything that points to our CMIP6 holdings. It is on my to-do list.
